src/commands/optimize.ts

```diff
@@ -0,0 +1,163 @@
+import { Flags } from '@oclif/core';
+import { Optimizer, Output, Report, ReportElement } from '@asyncapi/optimizer';
+import Command from '../base';
+import { ValidationError } from '../errors/validation-error';
+import { load } from '../models/SpecificationFile';
+import * as inquirer from 'inquirer';
+import chalk from 'chalk';
+import { promises } from 'fs';
+import { Example } from '@oclif/core/lib/interfaces';
+const { writeFile } = promises;
+
+export enum Optimizations {
+  REMOVE_COMPONENTS='remove-components',
+  REUSE_COMPONENTS='reuse-components',
+  MOVE_TO_COMPONETS='move-to-components'
+}
+
+export enum Outputs {
+  TERMINAL='terminal',
+  NEW_FILE='new-file',
+  OVERWRITE='overwrite'
+}
+export default class Optimize extends Command {
+  static description = 'optimize asyncapi specification file';
+  isInteractive = false;
+  optimizations?: Optimizations[];
+  outputMethod?: Outputs;
+
+  static examples: Example[] = [
+    'asyncapi optimize ./asyncapi.yaml',
+    'asyncapi optimize ./asyncapi.yaml --no-tty',
+    'asyncapi optimize ./asyncapi.yaml --optimization=remove-components,reuse-components,move-to-components --no-tty',
+    'asyncapi optimize ./asyncapi.yaml --optimization=remove-components,reuse-components,move-to-components --output=terminal --no-tty',
+  ];
+  
+  static flags = {
+    help: Flags.help({ char: 'h' }),
+    optimization: Flags.string({char: 'p', default: Object.values(Optimizations), options: Object.values(Optimizations), multiple: true, description: 'select the type of optimizations that you want to apply.'}),
+    output: Flags.string({char: 'o', default: Outputs.TERMINAL, options: Object.values(Outputs), description: 'select where you want the output.'}),
+    'no-tty': Flags.boolean({ description: 'do not use an interactive terminal', default: false }),
+  };
+
+  static args = [
+    { name: 'spec-file', description: 'spec path, url, or context-name', required: false },
+  ];
+
+  async run() {
+    const { args, flags } = await this.parse(Optimize); //NOSONAR
+    const filePath = args['spec-file'];
+    const specFile = await load(filePath);
+    this.isInteractive = !flags['no-tty'];
+    this.optimizations = flags.optimization as Optimizations[];
+    this.outputMethod = flags.output as Outputs;
+    const optimizer = new Optimizer(specFile.text());
+    const report = await optimizer.getReport();
```

Comment:
optimizer uses parser internally, so indirectly it also validates asyncapi documents. If this validation fails then `Error: There were errors validating the AsyncAPI document.` is thrown instead of `ValidationError`. I'm not 100% sure as when I look at your code I think you throw error properly.

we definitely need separate test coverage for it

Reply:
Since the asyncapi file is being validated before reaching the optimizer once, I don't see why it should fail when the optimizer parses it. I am not exactly sure how should I test it. 🤔

Reply:
> I don't see why it should fail when the optimizer parses it

I didn't get it

Reply:
sorry, my bad, I thought the asyncapi file is parsed by cli itself before getting to the optimizer. working on it right now. don't mind that comment.